src/deluge/gui/views/performance_session_view.cpp

```diff
@@ -0,0 +1,2045 @@
+
+		backupXMLDefaultFXPress[xDisplay].padPressHeld = FXPress[xDisplay].padPressHeld;
+		backupXMLDefaultFXPress[xDisplay].yDisplay = FXPress[xDisplay].yDisplay;
+		backupXMLDefaultFXPress[xDisplay].previousKnobPosition = FXPress[xDisplay].previousKnobPosition;
+	}
+	else {
+		//<status>
+		storageManager.writeTag(STRING_FOR_PERFORM_DEFAULTS_HOLD_STATUS_TAG, STRING_FOR_OFF);
+		//<row>
+		storageManager.writeTag(STRING_FOR_PERFORM_DEFAULTS_ROW_TAG, kNoSelection);
+		//<resetValue>
+		storageManager.writeTag(STRING_FOR_PERFORM_DEFAULTS_HOLD_RESETVALUE_TAG, kNoSelection);
+
+		backupXMLDefaultFXPress[xDisplay].padPressHeld = false;
+		backupXMLDefaultFXPress[xDisplay].yDisplay = kNoSelection;
+		backupXMLDefaultFXPress[xDisplay].previousKnobPosition = kNoSelection;
+	}
+
+	storageManager.writeClosingTag(STRING_FOR_PERFORM_DEFAULTS_HOLD_TAG);
+}
+
+//backup current layout, load saved layout, log layout change, update change status
+void PerformanceSessionView::loadPerformanceViewLayout() {
+	char modelStackMemory[MODEL_STACK_MAX_SIZE];
+	ModelStackWithThreeMainThings* modelStack = currentSong->setupModelStackWithSongAsTimelineCounter(modelStackMemory);
+
+	backupPerformanceLayout();
+	resetPerformanceView(modelStack);
+	readDefaultsFromFile();
+	logPerformanceLayoutChange();
+	updateLayoutChangeStatus();
+}
+
+//read defaults from XML
+void PerformanceSessionView::readDefaultsFromFile() {
+	FilePointer fp;
+	//PerformanceView.XML
+	bool success = storageManager.fileExists(STRING_FOR_PERFORM_DEFAULTS_XML, &fp);
+	if (!success) {
+		loadDefaultLayout();
+		return;
+	}
+
+	//<defaults>
+	int32_t error = storageManager.openXMLFile(&fp, STRING_FOR_PERFORM_DEFAULTS_TAG);
+	if (error) {
+		loadDefaultLayout();
+		return;
+	}
+
+	char const* tagName;
+	//step into the <defaultFXValues> tag
+	while (*(tagName = storageManager.readNextTagOrAttributeName())) {
+		if (!strcmp(tagName, STRING_FOR_PERFORM_DEFAULTS_FXVALUES_TAG)) {
+			readDefaultFXValuesFromFile();
+		}
+		storageManager.exitTag();
+	}
+
+	storageManager.closeFile();
+
+	if (!successfullyReadDefaultsFromFile) {
+		backupPerformanceLayout();
+		logPerformanceLayoutChange();
+	}
+
+	successfullyReadDefaultsFromFile = true;
+	uiNeedsRendering(this);
+}
+
+//if no XML file exists, load default layout (paramKind, paramID, xDisplay, yDisplay, rowColour, rowTailColour)
+void PerformanceSessionView::loadDefaultLayout() {
+	for (int32_t xDisplay = 0; xDisplay < kDisplayWidth; xDisplay++) {
+		memcpy(&layoutForPerformance[xDisplay], &defaultLayoutForPerformance[xDisplay], sizeParamsForPerformance);
+		memcpy(&backupLayoutForPerformance[xDisplay], &defaultLayoutForPerformance[xDisplay], sizeParamsForPerformance);
+		memcpy(&backupXMLDefaultLayoutForPerformance[xDisplay], &defaultLayoutForPerformance[xDisplay],
+		       sizeParamsForPerformance);
+	}
+}
+
+void PerformanceSessionView::readDefaultFXValuesFromFile() {
+	char const* tagName;
+	char tagNameFX[5];
+	tagNameFX[0] = 'F';
+	tagNameFX[1] = 'X';
```

Comment:
Should also have `tagNameFX[2] = 0` here, arrays are not auto-initialized to 0.

Reply:
> Should also have `tagNameFX[2] = 0` here, arrays are not auto-initialized to 0.

Is it needed if below I'm writing to &tagNameFX[2]? See line 1911.

Reply:
Right, I missed that before sorry.